server: refactor most of status endpoints to be provided over grpc with grpc-gateway for http

d4l3k opened the pull request:
This fixes most of #5530 by proxying requests over the standard grpc connection instead of the http admin interface on a different port.

Most of these changes are automatically generated protobuf output.

Converted endpoints:
- `/_status/details/{node_id}`
- `/health`
- `/_status/nodes`
- `/_status/nodes/{node_id}`
- `/_status/ranges/{node_id}`
- `/_status/gossip/{node_id}`

Things that remain as http and will be converted in a later PR:
- `/_status/logs/{node_id}`
- `/_status/logfiles/{node_id}`
- `/_status/logfiles/{node_id}/{file}`
- `/_status/metrics/{node_id}`
- `/_status/stacks/{node_id}` - Will probably remain as just HTTP.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6702)

<!-- Reviewable:end -->

d4l3k committed:
server: refactor most of status endpoints to be provided over grpc as well as over HTTP

This fixes most of #5530 by proxying requests over the standard grpc
connection instead of the http admin interface on a different port.

Most of these changes are automatically generated protobuf output.

Converted endpoints:
- /_status/details/{node_id}
- /health
- /_status/nodes
- /_status/nodes/{node_id}
- /_status/ranges/{node_id}
- /_status/gossip/{node_id}

Things that remain as http and will be converted in a later PR:
- /_status/logs/{node_id}
- /_status/logfiles/{node_id}
- /_status/logfiles/{node_id}/{file}
- /_status/metrics/{node_id}
- /_status/stacks/{node_id} - Will probably remain as just HTTP.

tamird commented:
General: try to avoid writing detailed PR descriptions - if you feel that this context is necessary, find a commit message to include it in. In this case, I think the PR description should be folded into the first commit message.

<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> #### server: refactor most of status endpoints to be provided over grpc with grpc\-gateway for http
> 
> This fixes most of #5530 by proxying requests over the standard grpc connection instead of the http admin interface on a different port.
> 
> 
> 
> Most of these changes are automatically generated protobuf output.
> 
> 
> 
> Converted endpoints:
> 
> * `/_status/details/{node_id}`
> 
> * `/health`
> 
> * `/_status/nodes`
> 
> * `/_status/nodes/{node_id}`
> 
> * `/_status/ranges/{node_id}`
> 
> * `/_status/gossip/{node_id}`
> 
> 
> 
> Things that remain as http and will be converted in a later PR:
> 
> * `/_status/logs/{node_id}`
> 
> * `/_status/logfiles/{node_id}`
> 
> * `/_status/logfiles/{node_id}/{file}`
> 
> * `/_status/metrics/{node_id}`
> 
> * `/_status/stacks/{node_id}` - Will probably remain as just HTTP.
> 
> 
> </details>


---

Reviewed 8 of 8 files at r1, 8 of 8 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[build/check-style.sh, line 100 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_q6k8RS5mhVVD_F0:-KHu_q6k8RS5mhVVD_F1:598923805) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d559ff0d768a46e1357672462edf52ecea98861/build/check-style.sh#L100)):*
> ```Shell
>   ! go tool vet . 2>&1 | \
>     grep -vE '^vet: cannot process directory .git' | \
>     grep -vE '^server/(admin|status)\..*\go:.+: constant [0-9]+ not a string in call to Errorf'
> ```

nit: this change should be in the first commit, not the second

---

*[build/protobuf.mk, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuXmHaMoC1KS4AJb92:-KHuXmHbb6StKjHIz2QH:904423887) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/build/protobuf.mk#L83)):*
> ```Makefile
> 	done
> 
> $(ADMIN_GW_GO) : $(ADMIN_PROTO) $(GO_PROTOS) $(GOGOPROTO_PROTO) $(PROTOC)
> ```

I believe we can avoid this duplication - come grab me and I'll show you what I mean.

---

*[gossip/gossip.go, line 551 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZCqRnR6QFzTfX47S:-KHuZCqRnR6QFzTfX47T:642007739) ([raw file](https://github.com/cockroachdb/cockroach/blob/929a323e7a4348a798e0722f6857c12fa0478fcd/gossip/gossip.go#L551)):*
> ```Go
> // GetInfosAsJSON returns the contents of the infostore, marshalled to
> // JSON.
> func (g *Gossip) GetInfoStatus() *InfoStatus {
> ```

why not return a value here?

---

*[gossip/gossip.proto, line 72 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZUFbbg9itSTs1kUz:-KHuZUFbbg9itSTs1kV-:708393955) ([raw file](https://github.com/cockroachdb/cockroach/blob/929a323e7a4348a798e0722f6857c12fa0478fcd/gossip/gossip.proto#L72)):*
> ```protobuf
> // InfoStatus contains information about the current status of the infoStore.
> message InfoStatus {
>   map<string, Info> infos = 1;
> ```

consider nullable = false

---

*[server/status.go, line 127 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuY0FQzVIIm3qdsRuk:-KHuY0FRQj5K4YWxKUBu:-1377565190) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L127)):*
> ```Go
> 	stores       *storage.Stores
> 
> 	// Mux provided by grpc-gateway to handle HTTP/gRPC proxying.
> ```

a structure like this one is now present both here and in server/admin - can we move hoist this stuff to server/server (I believe they can just be local variables, no need to store them on server.Server) so that there is only one instance of each?

---

*[server/status.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYR0hmtVvazVSF2n8:-KHuYR0hmtVvazVSF2n9:996479454) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L182)):*
> ```Go
> }
> 
> // RegisterGRPCGateway starts the gateway (i.e. reverse proxy) that proxies
> ```

would be good not to duplicate this code, as well.

---

*[server/status.go, line 216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZi25Gs56DsIDgIrh:-KHuZi25Gs56DsIDgIri:-938154490) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L216)):*
> ```Go
> // duplicates. This makes incoming requests to `foo/` go to the `foo` grpc
> // handler instead of `foo/{id}` and maintains previous compatibility with `foo//{id}`.
> type slashStripper struct {
> ```

this middleware is oddly specific - for one, it lives only on server/status, rather than server/server - is there a reason for that? It's also odd that it delegates straight to gwMux's ServeHTTP rather than server's.

---

*[server/status.go, line 341 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_FP5GBWf3xBWXXfk:-KHu_FP5GBWf3xBWXXfl:-1322711123) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L341)):*
> ```Go
> 
> // handleDetailsLocal handles local requests for node details.
> func (s *statusServer) handleDetailsLocal() *DetailsResponse {
> ```

nit: i think you can inline this into Details

---

*[server/status.go, line 617 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_Pf_az0eNIvL7ox2:-KHu_Pf_az0eNIvL7ox3:1984481895) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L617)):*
> ```Go
> 	rows := b.Results[0].Rows
> 
> 	resp := &NodesResponse{}
> ```

when you change NodesResponse.D to `nullable = false`, you should preallocate the slice here.

---

*[server/status.proto, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYexCmlV_zYxzNznc:-KHuYexCmlV_zYxzNznd:383990261) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L37)):*
> ```protobuf
>     (gogoproto.casttype) = "github.com/cockroachdb/cockroach/roachpb.NodeID"];
>   util.UnresolvedAddr address = 2 [(gogoproto.nullable) = false];
>   build.Info build_info = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "buildInfo"];
> ```

let's change the relevant UI code so we can get rid of this jsontag

---

*[server/status.proto, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYp2OtKd-LDvzhKkg:-KHuYp2OtKd-LDvzhKkh:-218118777) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L44)):*
> ```protobuf
> 
> message NodesResponse {
>   repeated status.NodeStatus d = 1;
> ```

nullable = false

also, how about a better name?

---

*[server/status.proto, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYqueOqop6td4QADB:-KHuYqueOqop6td4QADC:1694064763) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L52)):*
> ```protobuf
> 
> message RangeInfo {
>   roachpb.RangeDescriptor desc = 1;
> ```

should this be nullable?

---

*[server/status.proto, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYs7fEXdjvTsKlh14:-KHuYs7fEXdjvTsKlh15:-1265393466) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L61)):*
> ```protobuf
> 
> message RangesResponse {
>   repeated RangeInfo ranges = 1;
> ```

nullable = false

---

*[server/status_test.go, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZ-1Yb0Qyh3nTmzTz:-KHuZ-1Yb0Qyh3nTmzU-:1160644662) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status_test.go#L119)):*
> ```Go
> 		contentTypes := []string{util.JSONContentType, util.ProtoContentType, util.YAMLContentType}
> 		for i, contentType := range contentTypes {
> 			// Node status information isn't immediately available.
> ```

can you expand this comment with an explanation of the reason?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-:-KHuXYkiTsEhRikLRipg:-999456370)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[build/protobuf.mk, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuXmHaMoC1KS4AJb92:-KHv4yy8IiVqbf_rcOtg:1718019817) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/build/protobuf.mk#L83)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I believe we can avoid this duplication - come grab me and I'll show you what I mean.</details>
This amendment landed in the second commit, which is incorrect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 8 files at r1, 6 of 8 files at r2, 3 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[server/status.go, line 216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZi25Gs56DsIDgIrh:-KHv9Iz0LZtK2FYyis7G:956694665) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L216)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this middleware is oddly specific - for one, it lives only on server/status, rather than server/server - is there a reason for that? It's also odd that it delegates straight to gwMux's ServeHTTP rather than server's.</details>
Can we change the UI so this isn't necessary? We should be able to just remove the trailing slash from the urls used by the UI code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 3 of 16 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[build/check-style.sh, line 100 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_q6k8RS5mhVVD_F0:-KHuddaMNCeOoU6HwtYv:2008588587) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d559ff0d768a46e1357672462edf52ecea98861/build/check-style.sh#L100)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: this change should be in the first commit, not the second</details>
Is there any easy way to move changes between commits?

Just realized I folded in a bunch CR changes into the second commit.

---

*[build/protobuf.mk, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuXmHaMoC1KS4AJb92:-KHudpL_mU54bxf70-v9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/build/protobuf.mk#L83)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I believe we can avoid this duplication - come grab me and I'll show you what I mean.</details>
Done.

---

*[gossip/gossip.go, line 551 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZCqRnR6QFzTfX47S:-KHudrs2NvltDuXMfk6p:379403989) ([raw file](https://github.com/cockroachdb/cockroach/blob/929a323e7a4348a798e0722f6857c12fa0478fcd/gossip/gossip.go#L551)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not return a value here?</details>
You mean why *InfoStatus instead of InfoStatus?

GRPC expects a pointer to be returned and not by value. Seems simplest to have it be a pointer all the way through.

---

*[gossip/gossip.proto, line 72 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZUFbbg9itSTs1kUz:-KHufUI0vLy-yM89g6w9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/929a323e7a4348a798e0722f6857c12fa0478fcd/gossip/gossip.proto#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> consider nullable = false</details>
Done.

---

*[server/status.go, line 127 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuY0FQzVIIm3qdsRuk:-KHvCUdciOZB19rAA-6d:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L127)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> a structure like this one is now present both here and in server/admin - can we move hoist this stuff to server/server (I believe they can just be local variables, no need to store them on server.Server) so that there is only one instance of each?</details>
Done.

---

*[server/status.go, line 182 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYR0hmtVvazVSF2n8:-KHvCMVRzc3ks9yPlqVj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L182)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> would be good not to duplicate this code, as well.</details>
Done.

---

*[server/status.go, line 216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZi25Gs56DsIDgIrh:-KHuq13FLMItcxrLdOSq:1386794856) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L216)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Can we change the UI so this isn't necessary? We should be able to just remove the trailing slash from the urls used by the UI code.</details>
Updated PathForNodeStatus to not generate links like "/_status/nodes//5" and the UI package to hit "/_status/nodes" instead.

---

*[server/status.go, line 341 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_FP5GBWf3xBWXXfk:-KHugrXFeKAYg_7lc80v:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L341)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: i think you can inline this into Details</details>
Done.

---

*[server/status.go, line 617 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_Pf_az0eNIvL7ox2:-KHuhD5ZkH2ggapBiXNi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L617)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> when you change NodesResponse.D to `nullable = false`, you should preallocate the slice here.</details>
Done.

---

*[server/status.proto, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYexCmlV_zYxzNznc:-KHukM2_XdnsAH34z63P:-86524153) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> let's change the relevant UI code so we can get rid of this jsontag</details>
Turns out it's not used in the UI at all. 

I'm not entirely sure why it was named buildInfo in the first place. I can't find any other references to that name.

---

*[server/status.proto, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYp2OtKd-LDvzhKkg:-KHuhOAYoGtgU0ros3FK:1838430346) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nullable = false
> 
> also, how about a better name?</details>
That's what the existing code does.

https://github.com/cockroachdb/cockroach/blob/master/util/http.go#L150

I can change it to be 
```
repeated status.NodeData data = 1 [(gogoproto.jsontag) = "d"];
```
But that seems bad.

---

*[server/status.proto, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYqueOqop6td4QADB:-KHuiay9LROemfNDX46X:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should this be nullable?</details>
Done.

---

*[server/status.proto, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYs7fEXdjvTsKlh14:-KHuiWZe7w_ZTIUcD04U:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nullable = false</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 3 of 16 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[build/check-style.sh, line 100 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_q6k8RS5mhVVD_F0:-KHvPVpBHt0HOwIfoCyu:-1970126320) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d559ff0d768a46e1357672462edf52ecea98861/build/check-style.sh#L100)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Is there any easy way to move changes between commits?
> 
> Just realized I folded in a bunch CR changes into the second commit.</details>
Ignore that comment. Forgot to remove it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
I think I lost some changes to the proto files such as "gogoproto.nullable". Not entirely sure what happened with that.

tamird commented:
Reviewed 13 of 13 files at r5, 8 of 8 files at r6.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 551 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZCqRnR6QFzTfX47S:-KHvUuW-Myi6h1WeegyA:1170911349) ([raw file](https://github.com/cockroachdb/cockroach/blob/929a323e7a4348a798e0722f6857c12fa0478fcd/gossip/gossip.go#L551)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> You mean why *InfoStatus instead of InfoStatus?
> 
> GRPC expects a pointer to be returned and not by value. Seems simplest to have it be a pointer all the way through.</details>
I'd prefer a value - no need to put it on the heap until it's necessary - this method is quite far from the grpc use point.

---

*[server/server.go, line 551 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHvR70_FIIH03ULLDGw:-KHvR70a76FCY8w1pWom:506948623) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2abc1400e234206a116a192e0ff050f879231a/server/server.go#L551)):*
> ```Go
> 
> // Close closes all grpc-gateways.
> func (s *Server) Close() {
> ```

This doesn't appear to be used. I guess you moved it here because the admin server had a similar method, but you never hooked this one up to the stopper above.

Instead of this method, I suggest you just `stopper.AddCloser(stop.CloserFn(gwCancel))` above. I think with that change you'll be able to remove the gwCancel member from the server struct.

---

*[server/status.proto, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYexCmlV_zYxzNznc:-KHvUYm6l6cy_GxrbBAl:453675813) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L37)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Turns out it's not used in the UI at all. 
> 
> I'm not entirely sure why it was named buildInfo in the first place. I can't find any other references to that name.</details>
Is it really not used at all? Let's remove it then.

---

*[server/status.proto, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYp2OtKd-LDvzhKkg:-KHvUhDToT0Y0qFxRBni:631357423) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L44)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> That's what the existing code does.
> 
> https://github.com/cockroachdb/cockroach/blob/master/util/http.go#L150
> 
> I can change it to be 
> ```
> repeated status.NodeData data = 1 [(gogoproto.jsontag) = "d"];
> ```
> But that seems bad.</details>
I'd suggest changing the UI to match a more reasonable name. The existing code does that because it's a "generic" wrapper.

---

*[server/status.proto, line 36 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHvUeLW4IQ1MnQbxtwk:-KHvUeLW4IQ1MnQbxtwl:1656608970) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be3e1c9ec9b2ae0482a2c776a13d428337e8495/server/status.proto#L36)):*
> ```protobuf
> 
> message DetailsResponse {
>   int32 node_id = 1 [(gogoproto.customname) = "NodeID", (gogoproto.jsontag) = "nodeID",
> ```

how about this json tag, can we remove it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k committed:
Converted /_status/gossip/{node_id} to use grpc

d4l3k commented:
Review status: 1 of 17 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 551 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZCqRnR6QFzTfX47S:-KHva-OujBgfziFa8yWV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/929a323e7a4348a798e0722f6857c12fa0478fcd/gossip/gossip.go#L551)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'd prefer a value - no need to put it on the heap until it's necessary - this method is quite far from the grpc use point.</details>
Done.

---

*[server/server.go, line 551 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHvR70_FIIH03ULLDGw:-KHva8QCiHCAC8pdzXgn:995351734) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2abc1400e234206a116a192e0ff050f879231a/server/server.go#L551)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This doesn't appear to be used. I guess you moved it here because the admin server had a similar method, but you never hooked this one up to the stopper above.
> 
> Instead of this method, I suggest you just `stopper.AddCloser(stop.CloserFn(gwCancel))` above. I think with that change you'll be able to remove the gwCancel member from the server struct.</details>
I did add it to the stopper, but that makes more sense. Done.

---

*[server/status.proto, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYexCmlV_zYxzNznc:-KHvah8rhfqsKdg17ep7:-27524247) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Is it really not used at all? Let's remove it then.</details>
It's still useful information that people might want to have. Most of these endpoints aren't used by the UI.

---

*[server/status.proto, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuYp2OtKd-LDvzhKkg:-KHvbsr3pW-cJsGRYCXt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.proto#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'd suggest changing the UI to match a more reasonable name. The existing code does that because it's a "generic" wrapper.</details>
Done.

---

*[server/status.proto, line 36 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHvUeLW4IQ1MnQbxtwk:-KHvagYBCfvyxl9PZTY2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be3e1c9ec9b2ae0482a2c776a13d428337e8495/server/status.proto#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how about this json tag, can we remove it?</details>
Done.

---

*[server/status_test.go, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHuZ-1Yb0Qyh3nTmzTz:-KHzQLi8H41jyhRPtAdb:-171211965) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status_test.go#L119)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you expand this comment with an explanation of the reason?</details>
It's actually not needed. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod minor comments

<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> I think I lost some changes to the proto files such as "gogoproto.nullable". Not entirely sure what happened with that. One sec.</details>


---

Reviewed 16 of 16 files at r9, 8 of 8 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/node.go, line 57 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzRYlTogwhPJaIAwD-:-KHzRYlTogwhPJaIAwD0:1258347412) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9215256a408b8a17b9ce1b80b932ec57b45e0f3/cli/node.go#L57)):*
> ```Go
> 
> 	// Extract Node IDs from NodeStatuses.
> 	nodeStatuses := server.NodesResponse{}
> ```

Cool, we should port this cli command to grpc in a follow-on PR. See cli.runFreezeCluster

---

*[gossip/gossip.go, line 551 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzUxahuJBtHRH60cU6:-KHzUxahuJBtHRH60cU7:-1380682081) ([raw file](https://github.com/cockroachdb/cockroach/blob/003759775fe8a3e4c19f7c54bf7157cbd3ed9148/gossip/gossip.go#L551)):*
> ```Go
> 
> // GetInfoStatus returns the a copy of the contents of the infostore.
> // JSON.
> ```

remove this line

---

*[server/status.go, line 617 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_Pf_az0eNIvL7ox2:-KHzSbJv4EtkwvGAxhqX:552723102) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L617)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done.</details>
Can you actually preallocate it? `make([]status.NodeStatus, len(rows))` and then `row.ValueProto(&resp.Nodes[i])`.

also nit `resp` should not be a pointer - that will save an indirection during the iteration i suggested. not necessarily important, just wanted to call that out for future intuition.

---

*[server/status.go, line 639 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzT6GaxMBDWNhtC7l8:-KHzT6GaxMBDWNhtC7l9:1280490907) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9215256a408b8a17b9ce1b80b932ec57b45e0f3/server/status.go#L639)):*
> ```Go
> 	}
> 
> 	output := &RangesResponse{}
> ```

ditto - I'd prefer a value here because indirection. Also you could preallocate the `Ranges` slice, but this time use capacity since fetching the length isn't atomic with VisitStores.

---

*[server/status.go, line 284 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzUjnVyd394N0x3hjf:-KHzUjnVyd394N0x3hjg:-826653479) ([raw file](https://github.com/cockroachdb/cockroach/blob/003759775fe8a3e4c19f7c54bf7157cbd3ed9148/server/status.go#L284)):*
> ```Go
> 		return nil, err
> 	}
> 	return status.Gossip(ctx, req)
> ```

weird for this be called `status`. either rename this or the one above

---

*[ui/next/app/redux/nodes.ts, line 142 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzU2Ea4xAyWnqw7lf2:-KHzU2Ea4xAyWnqw7lf3:-2092944938) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9215256a408b8a17b9ce1b80b932ec57b45e0f3/ui/next/app/redux/nodes.ts#L142)):*
> ```XML
>     }).then((json: any) => {
>       // Extract the result, an array of NodeStatus objects.
>       let { nodes: jsonResult }: { nodes: any[] } = json;
> ```

we can do better than `any` now that these are protos. also, we need to regenerate the protobuf stuff in the ui. `make -C ui/next` aught to do it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-:-KHzV-QEG1Rm0sKPvdHp:-1971315572)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 8 of 21 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/node.go, line 57 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzRYlTogwhPJaIAwD-:-KHzdAsy6gEnFlxo7tVt:-480062378) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9215256a408b8a17b9ce1b80b932ec57b45e0f3/cli/node.go#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Cool, we should port this cli command to grpc in a follow-on PR. See cli.runFreezeCluster</details>
Yeah.

---

*[gossip/gossip.go, line 551 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzUxahuJBtHRH60cU6:-KHzaY57bEN-0YWPzCXy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/003759775fe8a3e4c19f7c54bf7157cbd3ed9148/gossip/gossip.go#L551)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove this line</details>
Done.

---

*[server/status.go, line 617 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHu_Pf_az0eNIvL7ox2:-KHzbZfNUPc9SVtJ3z1L:-1888021924) ([raw file](https://github.com/cockroachdb/cockroach/blob/92113eca3ee340742a9abcb3edffafc043582cd1/server/status.go#L617)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Can you actually preallocate it? `make([]status.NodeStatus, len(rows))` and then `row.ValueProto(&resp.Nodes[i])`.
> 
> also nit `resp` should not be a pointer - that will save an indirection during the iteration i suggested. not necessarily important, just wanted to call that out for future intuition.</details>
Done.

It's interesting that preallocating is the preferred style. At the team I was on last summer, every time I preallocated I got CR comments to just use append().

---

*[server/status.go, line 639 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzT6GaxMBDWNhtC7l8:-KHzcc3jPNBwoQ_K_ZRE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9215256a408b8a17b9ce1b80b932ec57b45e0f3/server/status.go#L639)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto - I'd prefer a value here because indirection. Also you could preallocate the `Ranges` slice, but this time use capacity since fetching the length isn't atomic with VisitStores.</details>
Done.

---

*[server/status.go, line 284 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzUjnVyd394N0x3hjf:-KHzb2F_i5q6RoY-9IHv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/003759775fe8a3e4c19f7c54bf7157cbd3ed9148/server/status.go#L284)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> weird for this be called `status`. either rename this or the one above</details>
Done.

---

*[ui/next/app/redux/nodes.ts, line 142 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KHzU2Ea4xAyWnqw7lf2:-KHzinrTmylu8JjdsPSS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9215256a408b8a17b9ce1b80b932ec57b45e0f3/ui/next/app/redux/nodes.ts#L142)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> we can do better than `any` now that these are protos. also, we need to regenerate the protobuf stuff in the ui. `make -C ui/next` aught to do it.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 13 files at r11, 8 of 8 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 18 of 18 files at r13, 9 of 9 files at r14.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[server/server.go, line 253 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI-NvG3liOMaFqlfx2m:-KI-NvG3liOMaFqlfx2n:1597908074) ([raw file](https://github.com/cockroachdb/cockroach/blob/c266233680711eccf1907e33757e209f6e062470/server/server.go#L253)):*
> ```Go
> }
> 
> // grpcGatewayServer represents a grpc service with HTTP endpoints through GRPC
> ```

fyi this comment is now unnecessary

---

*[server/status_test.go, line 46 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI-NWjPFLZ7MHM0jSvt:-KI-NWjPFLZ7MHM0jSvu:-1216644047) ([raw file](https://github.com/cockroachdb/cockroach/blob/c266233680711eccf1907e33757e209f6e062470/server/status_test.go#L46)):*
> ```Go
> )
> 
> // jsonpbUnmarshal unmarshals data using protoutil.JSONPb.
> ```

yikes, i don't think this is how i'd fix this test - will find you

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 12 of 25 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[server/server.go, line 253 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI-NvG3liOMaFqlfx2m:-KI3UBFWzhj4cWlIW64E:-2018578813) ([raw file](https://github.com/cockroachdb/cockroach/blob/c266233680711eccf1907e33757e209f6e062470/server/server.go#L253)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> fyi this comment is now unnecessary
</details>
But it doesn't hurt to have it right?

---

*[server/status_test.go, line 46 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI-NWjPFLZ7MHM0jSvt:-KI3U9cL1SA-dclviDdL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c266233680711eccf1907e33757e209f6e062470/server/status_test.go#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> yikes, i don't think this is how i'd fix this test - will find you
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 13 files at r15, 12 of 12 files at r16.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[acceptance/status_server_test.go, line 110 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_c9ki1KThnJ9TSdv:-KI3_c9ki1KThnJ9TSdw:292076375) ([raw file](https://github.com/cockroachdb/cockroach/blob/5430fae20fd2418260ed7fb0299b1aa1ae3af4bf/acceptance/status_server_test.go#L110)):*
> ```Go
> 	// Check local response for the every node.
> 	for i := 0; i < c.NumNodes(); i++ {
> 		checkNode(t, c, i, idMap[i], "local", idMap[i])
> ```

we need to add logic to checkNode to still hit the `/local` endpoint for each of `gossip,nodes,logfiles,logs,stacks`, when `nodeID == otherNodeID`, right?

---

*[acceptance/util_test.go, line 231 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_3kTWn93gddQnDHv:-KI3_3kUfxn4Zwz1xwdP:-2107162729) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/acceptance/util_test.go#L231)):*
> ```Go
> 
> // joinURL joins a base url and the relative path.
> func joinURL(base, rel string) (string, error) {
> ```

this is a convoluted version of `path.Join` - why not use that? You'll need to whitelist the linter, of course. Alternatively, can we make these tests not double-slash things?

---

*[server/server.go, line 253 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI-NvG3liOMaFqlfx2m:-KI3WtZG6uCLjJA8G-3F:-485110604) ([raw file](https://github.com/cockroachdb/cockroach/blob/c266233680711eccf1907e33757e209f6e062470/server/server.go#L253)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> But it doesn't hurt to have it right?
</details>
Sure.

---

*[server/status_test.go, line 123 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3XKdrB_8Zus-XQx3J:-KI3XKdsYREHAa_URZsH:-1473229687) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/server/status_test.go#L123)):*
> ```Go
> 				req, err := http.NewRequest("GET", s.Ctx.HTTPRequestScheme()+"://"+s.HTTPAddr()+spec.keyPrefix, nil)
> 				if err != nil {
> 					return err //t.Fatal(err)
> ```

probably didn't mean to keep these comments?

---

*[server/status_test.go, line 263 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_QHyUG58khSTHthX:-KI3_QHz3UjkoGADKQ7c:1331778940) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/server/status_test.go#L263)):*
> ```Go
> 
> // getRequest returns the results of a get request to the test server with
> // the given path.  It returns the contents of the body of the result.
> ```

double space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 10 of 25 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[acceptance/status_server_test.go, line 110 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_c9ki1KThnJ9TSdv:-KI3hNONjCiaX3C8LdQF:-502253829) ([raw file](https://github.com/cockroachdb/cockroach/blob/5430fae20fd2418260ed7fb0299b1aa1ae3af4bf/acceptance/status_server_test.go#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> we need to add logic to checkNode to still hit the `/local` endpoint for each of `gossip,nodes,logfiles,logs,stacks`, when `nodeID == otherNodeID`, right?
</details>
Added

---

*[acceptance/util_test.go, line 231 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_3kTWn93gddQnDHv:-KI3d54n1mKczCvjmE1_:-1096967872) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/acceptance/util_test.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is a convoluted version of `path.Join` - why not use that? You'll need to whitelist the linter, of course. Alternatively, can we make these tests not double-slash things?
</details>
`path.Join` strips out duplicate `/` so gives something like `https:/example.com/_status/details/local`

https://play.golang.org/p/wXIH_ShE3u

---

*[server/status_test.go, line 123 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3XKdrB_8Zus-XQx3J:-KI3e7NutTYtq5aJ4Htr:769088799) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/server/status_test.go#L123)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> probably didn't mean to keep these comments?
</details>
This was the area where status wasn't immediately available. I added them as a temporary measure

---

*[server/status_test.go, line 263 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_QHyUG58khSTHthX:-KI3eY862DJN83PWq0wX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/server/status_test.go#L263)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> double space
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 15 of 15 files at r17.
Review status: 13 of 25 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[acceptance/util_test.go, line 231 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_3kTWn93gddQnDHv:-KI3ia5dD6BXdTNuWiAi:-357443547) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/acceptance/util_test.go#L231)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> `path.Join` strips out duplicate `/` so gives something like `https:/example.com/_status/details/local`
> 
> https://play.golang.org/p/wXIH_ShE3u
</details>
Hm, and what does this method do?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 13 of 25 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[acceptance/util_test.go, line 231 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_3kTWn93gddQnDHv:-KI3sZHwe8cFXFWUTaWo:1692728226) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/acceptance/util_test.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Hm, and what does this method do?
</details>
This actually understands URL semantics so doesn't remove the `//` in `https://`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 12 of 12 files at r18.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[acceptance/util_test.go, line 231 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_3kTWn93gddQnDHv:-KI3uF2-sEdNKUvMBf_k:1725376038) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/acceptance/util_test.go#L231)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> This actually understands URL semantics so doesn't remove the `//` in `https://`.
</details>
Oh, I see. I suppose what I'm asking is: why was this function needed in places which were previously using concatenation? I presume the answer is double-slashing, but then why can't we fix those at the callers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 10 of 25 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[acceptance/util_test.go, line 231 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-KI3_3kTWn93gddQnDHv:-KI40K-0NxOrkZeGkQZk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/801467897871a05216a0566cb3014d219455280a/acceptance/util_test.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Oh, I see. I suppose what I'm asking is: why was this function needed in places which were previously using concatenation? I presume the answer is double-slashing, but then why can't we fix those at the callers?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> :lgtm: mod minor comments
</details>


---

Reviewed 3 of 15 files at r19, 12 of 12 files at r20.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6702#-:-KI43TEtBD6cgnN1vWeY:1417088721)*
<!-- Sent from Reviewable.io -->
